### ISSUE TITLE
Walk process algorithm

### ISSUE THREAD

**Issue by Eastrall** [Eastrall/Hellion#36]
The current algorithm is not calculating the mover positions very well. Which is a problem for the fight system regarding the monsters, because they need to check if they can attack a player.

I search the web for a new formula for the position update system and I came up with a new formula, but still, it's not working very well. We have almost the same result as the older system.

This is where the new algorithm is: https://github.com/Eastrall/Hellion/blob/develop/src/Hellion.World/Structures/Mover.cs#L325

Basically, the formula could be interpreted like this:

`currentPosition += normalized(destPosition - currentPosition) * speed`

In our implementation, we don't use the `Vector3` class, because it has the Y coordinate that we want to ignore. So everything is made by hand.
I believe that the problem is the `speed` value. Maybe we should look on this way.

If someone has an idea, feel free to post. Help wanted, thanks

**Comment by Eastrall** [comment]
I was doing some researches about this topic on the Internet, and I found out someone who asked something similiar on Unity3D forums. I was watching the solution, and in fact we don't use the `time` notion in our algorithm. Look at this example: https://docs.unity3d.com/ScriptReference/Vector3.MoveTowards.html

What is interesting here is this line: 
```c#
float step = speed * Time.deltaTime;
```
If we think about it, one `step` can be represented as a `distance`. And what is the formula for the distance? 
`distance = speed * time`

I think we might have a begining of an answer here! I'm gonna try it out later today! 

Notes:
`Time.deltaTime`: is the delta time between now and the last updated frame.

**Comment by Almewty** [comment]
After I got my server up and running I took a look at this.
It seems like the current position was kinda overshooting the destination position instead of stopping at the destination. Because of this it "rubberbanded" around the destination position (at least judging by the debug output).

In my branch I clamped it to the remaining distance if he would move more than he had to. See https://github.com/Almewty/Hellion/blob/walk-improvement/src/Hellion.World/Structures/Mover.cs#L358-L361

It seems to have improved the moving. If someone could try this change and verify the results it would be great.

**Comment by Eastrall** [comment]
I will try this solution later today, I'll give you a feedback.

**Comment by Eastrall** [comment]
I tried your solution. At the begining it wasn't working very well. But after that I noticed that one of the problem was the calculation of the `timeDelta` value. In fact the `timeDelta` was only calculated when the player was walking! So some times I would have a `timeDelta` equal to 10 seconds! Which is bad and the calculation were completly wrong.

What I've done is to move out the : 
```c#
timeDelta = Time.GetTick() - this.lastMoveTime;
this.lastMoveTime = Time.GetTick();
```

from the `WalkNew()` method to the begining of the `ProcessMoves()` methods. Now the time delta is always calculated. 
It seems that now the server is a little bit faster than the client .... Faster of maybe 0.5s - 1s depending where you click. One part of the problem is solved.

About your code, it's just perfect for the correction :wink:
